lib/client.js

```diff
@@ -1901,6 +1902,71 @@ MatrixClient.prototype.scrollback = function(room, limit, callback) {
     return defer.promise;
 };
 
+/**
+ * Take an EventContext, and back/forward-fill results.
+ *
+ * @param {module:models/event-context.EventContext} eventContext  context
+ *    object to be updated
+ * @param {boolean} backwards  true to fill backwards, false to go forwards
+ *
+ * @return {module:client.Promise} Resolves: updated EventContext object
+ * @return {module:http-api.MatrixError} Rejects: with an error response.
+ */
+MatrixClient.prototype.paginateEventContext = function(eventContext, backwards) {
+    // TODO: we should implement a backoff (as per scrollback()) to deal more
+    // nicely with HTTP errors.
+
+    var token = eventContext.getPaginateToken(backwards);
+    if (!token) {
+        // no more results.
+        throw new Error("No paginate token");
+    }
+
+    var dir = backwards ? 'b' : 'f';
+    var pendingRequest = eventContext._paginateRequests[dir];
+
+    if (pendingRequest) {
+        // already a request in progress - return the existing promise
+        return pendingRequest;
+    }
+
+    var path = utils.encodeUri(
+        "/rooms/$roomId/messages", {$roomId: eventContext.getEvent().getRoomId()}
+    );
+    var params = {
+        from: token,
+        limit: 30,
+        dir: dir
+    };
+
+    var self = this;
+    var promise =
+        self._http.authedRequest(undefined, "GET", path, params
+    ).then(function(res) {
+        var token = res.end;
+        if (res.chunk.length === 0) {
+            token = null;
+        } else {
+            var matrixEvents = utils.map(res.chunk, self.getEventMapper());
+            if (backwards) {
+                // eventContext expects the events in timeline order, but
+                // back-pagination returns them in reverse order.
+                matrixEvents.reverse();
+            }
+            eventContext.addEvents(matrixEvents, backwards);
+        }
+        eventContext.setPaginateToken(token, backwards);
+        return eventContext;
+    }).finally(function() {
+        eventContext._paginateRequests[dir] = null;
+    });
+    if (promise.isPending()) {
+        eventContext._paginateRequests[dir] = promise;
+    }
```

Comment:
I *think* this `isPending` check is redundant. All fulfilled/rejected handlers on Promises are set to go off asynchronously, so your `then()` and `finally()` blocks can't possibly have executed yet. If you were to immediately return a resolved/rejected promise (e.g. via `return q.reject("bad things");`) then this would be `false`, but I don't see why you'd ever want to do that. If there is a reason for that, please comment it as it is unusual.